File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    num_friends = len(friends)
    # name to index 변환 배열 n_idx
    n_idx = []
    for name in friends:
        n_idx.append(name)
    # 저번달 선물 개수 2차원 배열
    arr = [[0 for _ in range(num_friends)] for _ in range(num_friends)]
    
        
    # 이번달 받을 개수
    result = [0] * num_friends
    # 저번달 값 저장하기
    for gift in gifts:
        giver, receiver = gift.split()
        a = n_idx.index(giver)
        b = n_idx.index(receiver)
        arr[a][b] += 1
    
    # 선물 지수
    jisu = []
    for i in range(num_friends):
        val = sum(arr[i])
        for j in range(num_friends):
            val -= arr[j][i]
        jisu.append(val)    
    # 메인 로직
    for row in range(0, num_friends, 1):
        for col in range(row+1, num_friends, 1):
            n1 = arr[row][col]
            n2 = arr[col][row]                        
            # 1. 기록있음 
            if(n1 != 0  or n2 != 0):                
                if(n1 > n2):    
                    # n2가 n1한테 줌
                    result[row] += 1                    
                elif(n1 < n2):
                    # n1이 n2한테 줌
                    result[col] += 1                    
                else:
                    # 기록 같음
                    a_jisu = jisu[row]
                    b_jisu = jisu[col]                    
                    if(a_jisu > b_jisu):                        
                        result[row] += 1
                    elif(a_jisu < b_jisu):                        
                        result[col] += 1
            #2. 기록이 없음
            else:
                a_jisu = jisu[row]
                b_jisu = jisu[col]                
                if(a_jisu > b_jisu):                    
                    result[row] += 1
                elif(a_jisu < b_jisu):                    
                    result[col] += 1                                 
                
    # 결과 리턴    
    return max(result)
     

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
/**
friends : ["muzi", "ryan", "frodo", "neo"]
gifts : [
"muzi frodo",
"muzi frodo",
"ryan muzi",
"ryan muzi",
"ryan muzi", "frodo muzi", "frodo ryan", "neo muzi"]
*/
function solution(friends, gifts) {
  const N = friends.length;

  const friends_map = {};
  const board = Array.from(new Array(N), () => new Array(N).fill(0));
  const arr = new Array(N).fill(0);

  let count = 0;
  for (const friend of friends) {
    friends_map[friend] = count++;
  }
  console.log(friends_map);

  // 주고받은 선물 + 선물지수
  for (const gift of gifts) {
    const [from_name, to_name] = gift.split(" ");
    const from = friends_map[from_name];
    const to = friends_map[to_name];
    board[from][to] += 1;

    arr[from] += 1;
    arr[to] -= 1;
  }
  // console.log(arr);

  let cnt_arr = new Array(N).fill(0);
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (i >= j) continue;
      // console.log("testing, i: ", i, "j: ", j);

      const p1 = board[i][j];
      const p2 = board[j][i];

      // 주고받은 기록이 없거나 + 주고받은 수가 같다면 : 선물지수 큰 사람
      if ((p1 === 0 && p2 === 0) || p1 === p2) {
        // p1 받음
        if (arr[i] > arr[j]) {
          cnt_arr[i] += 1;
          // console.log("i가 받음");
        }
        // p2 받음
        else if (arr[i] < arr[j]) {
          cnt_arr[j] += 1;
          console.log("j가 받음");
        }
      }
      // 주고받은 기록이 있음 : 선물을 더 많이 준 사람
      else {
        // p1 받음
        if (p1 > p2) {
          cnt_arr[i] += 1;
          console.log("i가 받음");
        }
        // p2 받음
        else if (p1 < p2) {
          cnt_arr[j] += 1;
          console.log("j가 받음");
        }
      }
      console.log();
    }
  }
  console.log(cnt_arr);
  return Math.max(...cnt_arr);
}

